design/007-external-data-handling/007-external-data-handling.md

```diff
@@ -0,0 +1,561 @@
+accessible via Renku's `Dataset` API.
+
+Note: in the commands below `s3://` is used as an example but all of the support
+libraries referred to above support common APIs for (nearly) all likely storage
+backends and automatically switch between them depending on the protocol of the
+URL. That's a long way of saying that one can replace `s3://` with `hdfs://`,
+`nfs://` or `gcs://` etc.
+
+#### Creating a Renku dataset backed by external storage
+
+```
+renku dataset create s3-backed-dataset --storage s3://server/path/to/bucket
+```
+
+This command creates a dataset and sets its backing storage to be S3 instead of
+the default git-LFS. Locally, the directory `data/s3-backed-dataset` is created
+and added to `.gitignore`.
+
+```
+renku dataset add s3-backed-dataset /path/to/file/on/filesystem
+```
+
+This adds a file from the local filesystem to the dataset. It is _copied_ to
+`data/s3-backed-dataset`. (Should it also be pushed to the remote storage
+immediately?)
+
+#### Adding external data to a Renku Dataset
+
+This situation is different than the previous one in that remote data already
+exists but the user wants to bring it into a Renku context. The usual `renku
+dataset add` is used but Renku interprets the `s3://` URL and automatically
+registers it as an external dataset:
+
+```
+renku dataset add --create external-dataset s3://server/path/to/bucket
+-- prompt --
+Access Key: 1234
+Secret Key: abcd
+Store credentials? (y/N): y
+-- output --
+Credentials stored in ~/.renku/renku.ini
+```
+
+On the filesystem, the command creates `data/external-dataset` as is normal for Renku
+datasets. Importantly, `data/external-dataset` is also added to `.gitignore` to avoid
+exploding the git repository. 💥
+
+Note that by adding the data, only the metadata objects are created, the data is
+not actually staged yet.
+
+#### Getting information about the dataset:
+
+```
+renku dataset show external-dataset
+-- output --
+Name: external-dataset
+Created: 2022-03-30 15:55:59+02:00
+Creator(s): Rok Roškar <rok.roskar@sdsc.ethz.ch>
+Title: external-dataset
+Description: Data that is very large. Indeed.
+Data Sources:
+    - type: s3
+      url: s3://server/path/to/bucket
+      location: None
+```
+
+Note that `location: None` because the dataset has not been staged yet. See
+below for details on that.
+
+#### Staging external data:
+
+External data has to somehow be brought into the system where Renku is running.
+This seems to be a "staging" operation, but we could also call it something
+else. "Mounting" is another option but seems too linux-specific.
+
+```
+renku dataset stage external-dataset
+```
+
+This command copies by default. If the user wants to avoid copying and stage by
+mounting, FUSE can be used. For some filesystems we might be able to warn the
+user that the data they are trying to stage is very big and they should use the
+`--no-copy` flag instead.
+
+```
+renku dataset stage --no-copy external-dataset
```

Comment:
I think we should we be more explicit here, and make this flag `mount` or even `fuse-mount`, so it's clear what's happening (rather than what's _not_ happening)

Reply:
"fuse" is probably too jargon-y but maybe "mount" works?